bazel/SingleJar.java

```diff
@@ -0,0 +1,100 @@
+package grakn.buildtools;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class SingleJar {
+    private final static Pattern JAVA_PACKAGE_PATTERN = Pattern.compile("package\\s+([\\w.]+);");
+    private final static String SINGLEJAR_EXECUTABLE = "/tools/jdk/singlejar/singlejar";
+
+    private static String javaPackage(String sourcePath) throws IOException {
+        for (String line: Files.readAllLines(Paths.get(sourcePath))) {
+            Matcher lineMatcher = JAVA_PACKAGE_PATTERN.matcher(line);
+            if (lineMatcher.find()) {
+                return lineMatcher.group(1);
+            }
+        }
+        return null;
+    }
+
+    private static String unpackSingleJar() throws IOException {
+        File singlejar = File.createTempFile("singlejar", "");
+        singlejar.deleteOnExit();
+
+        if (!singlejar.setExecutable(true)) {
+            throw new RuntimeException("Could not make singlejar executable");
+        }
+
+        int i;
+        byte[] buf = new byte[8192];
+
+        try (InputStream fileStream = SingleJar.class.getResourceAsStream(SINGLEJAR_EXECUTABLE);
+             OutputStream out = new FileOutputStream(singlejar)) {
+
+            while ((i = fileStream.read(buf)) != -1) {
+                out.write(buf, 0, i);
+            }
+        }
+
+        return singlejar.getAbsolutePath();
+    }
+
+    private static void patchSourceJarLinks(String[] args) throws IOException {
+        boolean inResources = false;
+        for (int i = 0; i < args.length; i++) {
+            String line = args[i];
+            if (line.equals("--resources")) {
+                inResources = true;
+                continue;
+            } else if (inResources && line.startsWith("--")) {
+                break;
+            }
+
+            if (inResources) {
+                String[] fl = line.split(":");
+                if (fl.length == 2) {
+                    String originalFn = fl[0];
+                    String archiveFn = fl[1];
+
+                    if (originalFn.endsWith(".java") && archiveFn.endsWith(".java")) {
+                        // we are sure to map source file to source file
+                        String resultingPath = javaPackage(originalFn).replaceAll("\\.", "/") + "/" + Paths.get(originalFn).getFileName().toString();
+                        args[i] = String.format("%s:%s", originalFn, resultingPath);
+                    }
+                }
+            }
+        }
+    }
+
+
+    public static void main(String[] args) throws IOException, InterruptedException {
+        for (String arg : args) {
```

Comment:
what can `args` contain, @vmax ? can you give some examples?

Reply:
@lolski `args` are passed to `singlejar`. The part we are interested in is `--resources`, after which "A.java:B.java" statements go — which map source files on filesystem to their location inside JAR

Reply:
ok. is it in `source:destination` format?

ie., if `A.java` contains a class `grakn.core.A`, then are we changing the invocation from `A.java:A.java` to `A.java:grakn/core/A.java`?

Reply:
@lolski your understanding is correct here